celery/app/routes.py

```diff
@@ -22,34 +22,60 @@
 
 __all__ = ('MapRoute', 'Router', 'prepare')
 
+GLOB_PATTERNS = ('*', '?', '[', ']', '!')
+
+MAP_ROUTES_MUST_BE_A_DICTIONARY = """\
+Starting from Celery 5.1 the task_routes configuration must be a dictionary.
+Support for providing a list of router objects will be removed in 6.0.
+""".strip()
+
 
 class MapRoute:
     """Creates a router out of a :class:`dict`."""
 
     def __init__(self, map):
-        map = map.items() if isinstance(map, Mapping) else map
+        # map is either a mapping or a an iterable of tuples
+        if isinstance(map, Mapping):
+            map = map.items()
+        else:
+            warnings.warn(
+                CPendingDeprecationWarning(MAP_ROUTES_MUST_BE_A_DICTIONARY))
         self.map = {}
-        self.patterns = OrderedDict()
+        patterns = {}
         for k, v in map:
             if isinstance(k, Pattern):
-                self.patterns[k] = v
-            elif '*' in k:
-                self.patterns[re.compile(fnmatch.translate(k))] = v
+                # This is already a regular expression so we simply store it.
+                patterns[k] = v
+            elif any(glob_pattern in k for glob_pattern in GLOB_PATTERNS):
+                # This is a glob pattern so we
+                # must to translate it into a regular expression.
+                patterns[re.compile(fnmatch.translate(k))] = v
             else:
+                # This is a direct mapping between a task and a routing
+                # so we simply store it.
                 self.map[k] = v
 
+        # We sort by the regex pattern's length since longer regex patterns
+        # are likely to be more specific.
+        self.patterns = tuple(reversed(sorted(patterns.items(),
+                                              key=lambda item: len(item[0].pattern))))
+
```

Comment:
**[suggestion]**: Treat each type of pattern different.

After trying out some examples manually in `ipython` I suggest giving different weights to each of the type of patterns we can receive from the user.

We can receive either a regex, a glob pattern or a direct mapping.

We could keep the `OrderedDict` but insert direct mapping first, then glob patterns then regular expressions. Keeping the length sorting. This way we avoid any weird cases where a string is actually longer but is not more specific than a direct mapping.

Something like this:

```python
    mappings = []
    globs = []
    regexes = []

    for k, v in map:
        if isinstance(k, Patter):
            regexes.append((k, v))
        elif any(glob_pattern in k for glob_pattern in GLOB_PATTERNS):
            globs.append(re.compile(fnmatch.translate(k)),  v)
        else:
            mappings.append((k, v))

    globs = reversed(sorted(globs, key=lambda item: len(item[0].pattern)))
    mappings = reversed(sorted(mappings, key=lambda item: len(item[0].pattern)))

    self.patterns = OrderedDict(mappings + globs + regexes)
```

This way any direct mapping will always take precedence over everything else, then glob patterns then regexes. 
This is mainly because in my opinion `some.direct.mapping` is more specific than `some.direct.mapping.*`

Then, we only sort direct mappings and glob patterns by length because that does make sense but we leave regexes in the same order we receive them. That way we keep the same behavior for regexes that's used in most frameworks using regex to match stuff, the first one wins.

Reply:
if isinstance(k, Patter): should be, if isinstance(k, Pattern):

Reply:
and  self.patterns = OrderedDict() to  {}

Reply:
Why does the order of the regexes matter?

Reply:
Also, I'd prefer to have consistent behavior across all route types.

Reply:
The order of the regexes matter because that's usually how regexes are matched. This is mainly thinking about other python web frameworks that accept regexes as url routes. But, it might make more sense for urls and not here.

I'm OK with also sorting the regexes by length. But I do think that the precedence should be direct mappings > glob patterns > regexes.

It makes more sense from the user perspective that a direct mapping would work over a glob pattern or a regex since a direct mapping is not leaving room for any other value and is an exact string match.

Reply:
In the current code (even after this change), we prefer direct mappings over patterns.
What I don't understand is why we should prefer glob patterns over other regexes.

Reply:
I'd assume we want to give preference to glob patterns over regexes because glob patterns are more specific by definition.

For example `using.glob.patterns.*` is more specific than `using\.(glob|regexes)\.patterns\..*`.
This is a very specific example but since we cannot know how specific or general a regex is then I think it's safe to assume that a user would expect a string with a single glob pattern will have precedence than a regex with a bunch of rules.

And, you are correct I forgot about the code in `__call__` that checks for direct mappings before looping through the rest of the configured routing patterns.

Reply:
If it has a bunch of rules then it's longer by definition no?

Reply:
Yes. The code in this PR would prefer the longer regex, right?
In this case, the longer regex is not more specific than the sorter pattern with a glob.

Or, what do you mean?

Reply:
Why is it not more specific?

Reply:
I think we could arguably sort patterns by the number of groups in them, since that likely how most users would be using them. Aside from that, I think the behaviour as described in this comment thread seems sensible enough, and we should just aim to document it in a fairly obvious way so that if/when it surprises someone they can find out why it's not doing what they think it should.